### DISCUSSION TITLE
ideas to control Ableton's Browser via MidiRemoteScripts

### DISCUSSION THREAD

**Discussion by raphaelquast** [raphaelquast/beatstep#15] [Ideas] [answered]
## 📁 QBrowser 
-  implemented in Beatstep_Q > v1.8 !  🥳

**Comment by Zedrock** [comment]
Hello,
I wanted to thank you for your hard work which transformed my little beatstep into a very complete controller for Ableton.

All it needs to be complete could be the Ableton Browser navigation ... and above all the possibility of selecting an item (*.agv, *.adv, etc) in this famous browser.

I saw that you had thought of implementing this navigation with a function : def _toggle_browse in QControlComponent.py.

I imagine that, like many others, you must have realized that the selection of an object in this menu was - obscure - , maybe not possible as a - no way.
It seems that the lack of a kind of mapping for the ENTER/RETURN key has a lot to do with it.

After much trial and error, I finally found a way around this.

- We can't use real python libraries directly from Ableton MIDI Remote Scripts folder BUT we can call an external executable from the MIDI Remote Scripts.

I will try to explain how you could do it :
- With Python27 installed
- Install keyboard package under python27
- Install auto-py-to-exe package to make our send_enter_key.py as send_enter_key.exe
- Make a simple python script send_enter_key.py as follow : 
``
#!/usr/bin/env python
``
``
#coding=utf-8
``
``
from __future__ import with_statement
``
``
import keyboard
``
``
#Press ENTER
``
``
keyboard.send("enter")
``
-  This script send the ENTER keystroke to the current windows application.

- Then, execute auto-py-to-exe
- Once done, just copy the ouput folder named send_enter_key (with send_enter_key.exe and all needed dependencies in) 
to C:\ProgramData\Ableton\Live 10 Suite\Resources\MIDI Remote Scripts\Beatstep_Q

Now, from QControlComponent.py, you can make a call to send_enter_key.exe as you need with : 

``
import subprocess
``
``
subprocess.Popen([C:\ProgramData\Ableton\Live 10 Suite\Resources\MIDI Remote Scripts\Beatstep_Q\send_enter_key\send_enter_key.exe])
``

It works with Ableton LIVE 10, under Windows 7 (64b).

Maybe this could help you to ending the Browser control part.

Best regards

Zed'

* Sorry for my english, this is not my native speaking.

**Reply by raphaelquast** [reply]
Hello!
Great to hear that you like what I did so far! 🚀 

Your are right... controling the browser would be really nice, and a while ago I've tried to include browser-control feature, but the lack of documentation and the missing callbacks for easy interaction made me focus on getting the other parts running first...

I think the main problem with the browser-control is that there are many possible objects that one can select.... so this feature must be carefully implemented to avoid messing up your project by opening something you never wanted to open...

Your method of compiling an external script is a really nice idea for a workaround, but there are some drawbacks when thinking about distributing it to the public:

- I'm not 100% sure, but an explicit call to an external .exe from the script might rise some security concerns 
  and also issues with antivirus warnings etc.
- having to compile a script might be "out of scope" for many users and I'd like to keep the usability as simple as possible
   (also one would have to think about how this works on different operating-systems...
   - ... also adding a dependency to a non-native library like `keypress` would require me to keep track of its development
     (right now it's only native python and the Ableton API, which keeps maintainance on this part minimal)
- using system-wide keypress events might interfere with other software as well (e.g. if you change the focus)
- from Ableton 11 on, MidiRemoteScripts run on python 3   
  (but that's no real concern since your method should be right away portable to python 3)

There's even a python-native way for simulating keypress-events (using [cpython](https://docs.python.org/3.7/library/ctypes.html) to call windows "user32.dll" functions) but that's still windows-only...

Right now I'm working on finalizing my implementation of a "midi sequence editor" (#42) functionality (it's awesome, but unfortunately exclusive for Ableton 11 since they changed the API for Midi notes quite a bit), but knowing that there's some demand on a browser-control feature, I might give it another try as well...

I do think that it should be possible to add a basic browser control functionality with the existing API  
(if you can do it with Ableton Push, BeatStep_Q should be able to do it as well 😄 )

I've opened a dedicated issue (#45) to keep track of this but it'll take a while until I find the time to actually work on it...
If you're interested in contributing to the implementation, I'm happy to discuss how (and what) to implement or to have a look at pull-requests.

Greets, Raphael

... by the way, you're english is perfect (but I'm no native speaker as well)

**Reply by Zedrock** [reply]
Hi,
Huhu^^ Really, BeatstepQ is an amazing idea. It gave my beatstep a second life ;) 
Ho! And your overlay is just excellent! I brainstormed all over the place to come up with one. Very smart the way you design yours with the colored - pipe - lines! Congratulation.

I understand your remarks about the solution I thought of for the ableton's browser.
- Indeed, I do not have enough knowledge regarding the safety of the process.
- Yes, the - focus - on the right window application is still a problem... the fix for that, working only for Windows system, is to use same mechanism as used in AutoHotkey (with pywin32 packages for example) to play with Foreground and Background windows.
- There are many others python packages for sending - virtual - keystroke. I choose the old Keyboard package because it was very simple  to use to make some test. And this old package is cross-platform.
- Ableton LIVE has implemented its own python, which is almost the same as the common installation but alas, some libraries are missing or not fully implemented (in LIVE 9 or 10), notably the ctype library which would have been very useful to create a procedure for sending a virtual key.

- By now, many people seems to use Bome MIDI Translator or AutoHotkey to map MIDI controls and Keystrokes. This is not really a solution that suits to me.

- We have an idea about how Push deals with the browser. I worked myself on Push's python script mechanism in order to reproduce it. Someone pretty well summarized how push works on the ableton forum : [How to : use midi remote script Major Update (sparkle)](https://forum.ableton.com/viewtopic.php?f=1&t=200513&start=30#p1606429)

Here's an extract : 
> [...] I have been unable to find a way to get it working (the only way is to keep the enter key of your keyboard next to the sparkle). I have taken a look to the push broswer and their is a major difference: as the push controller have a screen, you can import the module " information" such as name etc... and create a browser. With the sparkle, the only way i had found was to control the Live software and to select an instrument or another. if we had a screen i would have done a browser [...]

- One other way would be to see with Max For Live.

I have some other pythonic ideas about how to control the Browser... I will continue my attempts and I will keep you informed if I ever find other working solutions.

Thanks again for your attention and for your MIDI remote script and its essential overlay.

Best regards

Zed'

**Reply by raphaelquast** [reply]
Hey, 
thanks! it's good to see that people start using it and appreciate the work I put into it!
(for me this whole project is one of the good things that came out of the pandemic so far 😅 )
It took me quite some time (and a bunch of different concepts) to come up with the overlay design...  
At one stage it simply became necessary to make one since remembering all the assignments is no longer an option 😄 

- i agree, using a Midi-2-Key translator could also be a workaround (and as probably most people who work with MidiRemoteScripts I spent my hours with those tools as well... 😄) However I too find it's a rather cumbersome way of doing things and having all up and running when the controller gets plugged-in was one of my primary motivations for starting the BeatStepQ script.

- thanks for the link to the discussion about how Push's script is dealing with the browser! It'll certainly be helpful when I try to dig into it myself... (even tough it seems that most people got stuck while trying...)

- concerning Max4Live, there are 2 main reasons that kept me from using it so far:
   - first, i don't really know hot to use it so I guess it would take quite some time to learn
   - second, and most importantly... it's not a free software and I do not own a license   
      (i know it's free for running scripts but as far as i know not for developing them)


Any help with the feature (no matter if it's just ideas or actual code) is highly appreciated! 

Cheers, Raphael

**Reply by raphaelquast** [reply]
hey, I gave this Browser-control idea a bit more thought, and the only thing I'm actually missing to implement a basic browser functionality is a way to get the currently selected `BrowserItem`.

I can scroll the browser and get a list of all possible items, but somehow the property `BrowserItem.is_selected` that is supposed to do the job (from the Ableton API) always returns `False` so I have no clue how to get the selected item....

once the item is identified, you can use `app.browser.load_item(item)`
to load instruments, effects or swap presets. (without the need for a simulated enter-button)

@Zedrock do you by any chance have an idea on how to access to the currently selected item from the browser?

Cheers, Raphael

**Reply by Zedrock** [reply]
Hello Raphael,

- Yes, the BrowserItem.is_selected and Browser.load_item(item) are the things to resolve.
That's the mystery : we can open/close the Browser, we can navigate throught it. But we can't validate the selected item.
We are not able to know internally (with python script) "where" and "on which item" in the menu "index" we are. What is needed to finally load the choosen item.

- One first solution I tried was to **"mimic" the browser** navigation by python script, not necessary synchronized with the real browser. Alas, the big lack with it is to not have efficient visual feedback (except by a show_message) in Ableton, to know on which instrument/preset we are.
That's similary to how Push/Push2 script deals with it. The browser is "replicated" with a complex Model/View scripting then the feedback is displayed on the hardware screen (without any synchronized feedback in the Ableton browser).

- After some try, **BrowserItem.is_selected is working** (False/True) when an instrument/preset is previously loaded on track before, related to the hotswap flag.
This let me think that a way to explore could be to use hotswap flag and filter_type property.

- An other way to explore : **the URI property** of the current item. By now, I don't know how to use it to access/call the current item from the browser database. There surely a reason why we have access to URI property.
(This URI seems to be related to the QML mechanism which is used to render the browser. With a very deep retro-engineering, maybe we could use the URI property of the browser's item to select it. By now, I have no idea how to access the QML tree from python script and may be a wrong way).

- That's why, in my mind, the simulated enter-button was the more simple solution to have acces to the selected item shortcut (which is the main lack).

This is a short summary about where I am with the browser by now.

Best regards,

Zed'

**Reply by raphaelquast** [reply]
Hey, thanks for your insights!

I guess one reason why there's no proper browser-binding is that the browser can do more than just selecting a single item... so a binding to the python API would have to be quite extensive... (e.g. multiple selections etc..)

- You are right concerning the `is_selected` property... it is connected to the hotswap-feature and only true if hotswapping is active... 
  (also it sticks to the target being hotswapped... not to the item that is selected in the browser... so it does not help us at all)

- from what I see, the **URI property** will also not help... it's just a string-identifier of a given browser-item (e.g. something like *"query:Synths#Drum%20Rack"*  for the DrumRack)  but to use it you already need a reference to the object you're interested in... 


one additional option would be to have a "floating-window" browser that is connected to the control-surface
(and completely ignore the built-in browser tab)
right now I have no idea how i could achieve this without using Max... but it is certainly possible if you have a dedicated midi device that uses a floating-window...   (e.g. something similar to this one could do the job: [Set List / Text Display" Max For Live device ](https://youtu.be/Q9fjOHT4dIM?t=181) )

...Of course python-native GUI modules like tkinter are not shipped  with the python-distribution 🤯 so a simple python-gui window is not possible...


however this idea seems even more chumbersome than pre-compiling a script that mimics a keypress-event....
I guess I've dismissed your enter-button idea too quickly 😅 
It really seems to be the only option that allows having visual feedback via the browser...

Cheers, Raphael

**Reply by Zedrock** [reply]
Hi,

- From what I know after multiple searches, no script gives access to an interactive selection of instruments/prests with the Ableton Browser. Even more elaborate scripts like **ClyphX Pro** from Isotonik, or **Remotify**, do not go beyond: opening / closing the browser. (And now, we know why ;) )

- I found an Ableton python script for the **Maschine Jam**, https://haxorroxor.com/post/186562511605/hxr-jam-for-live-10, which prepares as many buttons as necessary on the Jam to distribute and launch the audio effects and midi effects (related to the current selected track type) to select instrument (the *.adv only, not the *.adg presets).
This is the same solution offered by the IOs **touchABLE Pro** tablet application. Still no interaction and feedback with the Ableton Browser.

- I'm afraid that if even professionals haven't implemented real-time browsing in Ableton (and, the item selection), it won't be possible from the API.

- I retain your idea of a MAX display interface. I also started to explore this route but I had not yet finalized how to display the information that can be used in this way.

... to be continued^^.

**Reply by raphaelquast** [reply]
Hey, so again I've spent some time (this time actually quite a bit 😄) 
- I've checked your "enter-button" idea but unfortunately Ableton 11 is missing the `msvcrt` module (a dependency of `subprocess` on windows) so there's no chance for this to work at all (there's still the not-recommended `os.system()` function to execute a system-call but as far as i know it will not forward any inputs to external applications...)

- after finding the above issue, I've decided to go a completely new way... behold... the Ableton status-bar Browser 🥳 !

To be more clear... I've decided to implement a basic browser-functionality that works with the status-bar.
It's only a 1-line output, so not recommended for searching deep within the library, but it nicely serves the purpose of loading things quickly if you know where to find them.  

... here's a gif that shows how it currently looks like:
- on the left you see the parent-folder hirachy of the selected item
- a red circle before the selected item indicates that the item can be loaded
![QBrowser](https://user-images.githubusercontent.com/22773387/114905924-cc117800-9e19-11eb-8ec8-743537a28460.gif)


So far I have implemented all features that are needed for a basic browser... 
If you're interested in having a look, all details are here:  #60 .
(the feature is developed on the "browser" branch)
- It should be compatible with ableton 10... (however without the fancy symbols)

Let me know what you think! (or if you have some ideas for improvements)!

**Reply by Zedrock** [reply]
Hey Raph',
This status bar feedback is so really usefull in so many cases (for the sequencer, as exemple).

That one looks great! Good work!
A fine and simple solution without any dependencies I think!

I can't wait to check it ;)

Thanks a lot to have done a look deeper on this damn' browser^^
I hope it will prove its value over time.

Zed'

**Reply by raphaelquast** [reply]
hey, 
it's already fully functional for Ableton 10 & 11 (just some feature-ideas missing as indicated in #60 )
I already played around a bit with it and I think it is a really awesome extension to the BeatStep_Q script!!
- now you can finally sit on the couch with a laptop + the BeatStep and have full   
   control over ableton without having to fiddle around with a mouse all the time 😄 

if you'd like to give it a try, just go here [>> browser-branch](https://github.com/raphaelquast/beatstep/tree/browser) and replace the files in your MidiRemoteScripts folder!

some comments on the functionalities, button-assignments etc. would be very welcome!

the biggest drawback right now is that the unicode-symbols do not work in ableton10 ... so I need to search for a set of symbols that do the job equally well as the ones selected for ableton 11 (aside of that all should work as expected)

**Reply by raphaelquast** [reply]
hey, good news 🥳 
I was able to fix the unicode-symbol issue with python2... now the QBrowser feature is fully compatible with Ableton 10!

**Reply by Zedrock** [reply]
Hi,

And it works fine! Well done!
Formating the status bar text as you did is just so smart!

By now, I have not any special comment on the functionnalities or button-assignement. After a short time to learn, understand and pratice it then QBrowser does the essential tasks I need!

I am still very impressed by the way you gave the beatstep so many features with only 3 main buttons and the shift one!!!
We can even imagine adding more functions in the future by playing on the 16 buttons and their pressure time ^^

Now, for my common use, BeatStep Q is not so far from being mainly (and much more) completed! ;)

Some minor remarks : 
- With Beatstep Q, by now, I'm used to navigate through scenes and tracks with the encoders 8 and 16. Is this meaning that buttons 8 and 16 could be assigned to other functions, instead tracks or scenes navigation, if needed, later ?

- Is there any reason why you have not added Max For Live, Plug-ins and Clips folders in QBrowser ?

Ho! A last thing : 
- Do you think remote script could be able to add a missing beatstep function : patterns chain ?

Best regards

Zed'

**Reply by raphaelquast** [reply]
Hey, 
Thanks! It was (is) a lot of work but it gave my dusty beatstep a completely new life.
At one stage I simply wanted to try how much I can get out of this little controller 😄 

In theory there is no limit to the functionality of the script... but I think with the browser-feature it kind
of reached it's complexity limit (I guess more layers would no longer be user-friendly...)
Also I aggree that with the addition of the QBrowser, the toolset is more or less complete...

After trying (e.g. failing... ) several different approaces with "enter-button" functionalities and hacky ways to obtain
the currently selected item in the browser, the info-bar seemed the only possible way to have something that is   
usable without requiring a complicated setup.    
First I wasn't sure if the usability would be good enough (after all it's just a single line of text)  
but now I'm really happy how it turned out!



- concerning buttons 8 & 16:
   yep, they could be mapped to do other things... however, I do like having a button to navigate
   through clips and tracks since the encoders are sometimes a bit jumpy...
   however, If there's a good reason (e.g. a nice feature-idea), they might be re-assigned in the future.

- I've added "Max For Live", "Plug-ins" and "Clips" folders now!
  ...they are not accessible via the quick-select (buttons 1-7)  but they appear in the top-folder-level and 
  can be browsed just as any other folder.


I don't understand what you mean by your "last thing"... could you clarify what you mean by "patterns chain" ?

**Reply by Zedrock** [reply]
Sure,

When I'm playing my pattern 1, the beatstep is playing and loop on the pattern 1.
Do you think if it could be possible, with python script, to make it play pattern 2 once the pattern 1 is played. A kind of chain between pattern 1 and 2 (without needed to press recall pattern 2 to switch from pattern 1 to pattern 2).
And so on if I want to play pattern 1 then 2 then 3 then etc...
If this function already exist on beatstep, I never found it, however this is a common one on many sequencer.

I hope it's more clear.

Zed'

**Reply by raphaelquast** [reply]
Hey, it's built into Ableton, but unfortunately it is not accessible via MidiRemoteScripts.
-> [follow actions](https://youtu.be/3n1hahB1Y7E?t=93)

You simply need to activate it once on a clip and then you can copy the clip and they will trigger each other.

**Comment by raphaelquast** [comment, marked as answer]
Hey, I'll be marking this discussion as "answered" since the Browser has come to reality already 🥳
